### ISSUE TITLE
[css-inline-3] initial-letters: interaction of shape-margin and regular margin

### ISSUE THREAD

**Issue by faceless2** [w3c/csswg-drafts#5119]
@fantasai as you're working on these.
```css
p::first-letter {
    initial-letters: 3;
    margin-right: 20px;
    shape-margin: 10px;
    initial-letters-wrap: all;
}
```
Do we add the 20px margin to whatever the 10px margin around the shape is calculated to be at that point, or do we use the maximum of the two values?

We've implemented using the maximum of the two margins, which seems to work better - you can set `shape-margin` to ensure the shape isn't adhered to too closely, and `margin-right` to push the content along. This also makes more sense with `initial-letters-wrap: first`, where both properties apply to the first line but only `margin-right` on subsequent lines.

**Comment by fantasai** [comment]
Maximum of the two values. The content-box size (and consequently the margin box size) is not influenced by `shape-margin`.

**Comment by fantasai** [comment]
Actually, thinking about this more... we've got several options here:

- Lines wrapped to the glyph use the `shape-margin` distance between the glyph outline and the line box edge, completely ignoring margin/border/padding on the initial letter box itself.
- Lines wrapped to the glyph use the maximum of the `shape-margin` distance and margin+border+padding distance between the glyph outline and the line box edge.
- Add a `none` initial value to `shape-margin`, and use `shape-margin` exclusively whenever it is not `none`.

Only honoring `shape-margin` rather than taking the maximum is simpler, but there is a definite advantage to honoring the margin/border/padding distances, since they can be controlled independently per axis. And anyone applying either margin or padding automatically gets some amount of spacing on all lines, and does not need to specify both `shape-margin` and `margin-inline-end` to get reasonable behavior. So I think I'm a bit biased against the first option, but either of the second two seem reasonable.

(The additive option, having `shape-outline` expand the size of the initial letter box, doesn't seem particularly useful since `padding` accomplishes the same thing; and `shape-margin` in general overlays the box decorations rather than affecting the box’s size, so that would be inconsistent with how it's interpreted for floats also.)

CC @dauwhe

**Comment by astearns** [comment]
I think it's better to have an interaction between the properties rather than a switch where we only use one or the other.

I think using the maximum is reasonable. Another possibility is to make things consistent with how shapes and margin boxes in floats interact - the shape described by shape-outside and shape-margin is clipped to the margin box. It's a bit weird for authoring, but once you learn how things are implemented for floats your knowledge would carry over to initial-letter.

**Comment by faceless2** [comment]
Hmm. Actually I prefer Alan's suggestion. My initial case for adding `shape-margin` was to stop this sort of situation:
![image](https://user-images.githubusercontent.com/989243/84881429-c0341880-b085-11ea-93cd-cb42d6be04b2.png)

But setting`shape-margin` will add space around the glyph in all directions, pushing the text below the `initial-letter` down by an additional line. Clipping the shape to the margin box means the user can prevent this by leaving `margin-bottom: 0` a the default value of 0.

**Comment by faceless2** [comment]
We've tried implementing Alan's suggested approach, and there's a slight problem. The spec states that when using `initial-letters-wrap: all`, the outline is expanded by the end-side margin+padding+border. This means we can't effectively use the margin box to clip the shape:

Consider this:
```css
p::first-letter {
    initial-letter: 3 3;
    initial-letter-wrap: all;
    shape-outside: 10px;
    margin-right: 20px;
}
```
![image](https://user-images.githubusercontent.com/989243/84899554-b7047500-b0a0-11ea-8c90-81e4d6fde4b3.png)

The black "G" is the actual letter. The `margin-right` means we expand the outline 20px to the right, as shown by the green letter. It also expands the margin-box, as shown by the red outline. The 10px `shape-outside` is clipped by the margin-box, but it doesn't have any effect on the overall outline - thanks to that green letter G.

I think the best solution would be to not add the end margin+padding+border to the shape - i.e. remove "_plus the amount of the initial letter’s end-side border+padding+margin._" from the definiiton of `initial-letters-wrap: all`.

In other words: the margin sets the margin-box, and does nothing else. The letter shape is expanded by `shape-margin` alone, and that shape is clipped to the margin-box. So it works exactly like it does on floats.

This would be a change to the current functionality, but - as Webkit hasn't implemented `initial-letters-wrap: all`, and the only other implementations of `initial-letter` are us and @stantonma, you'd probably get away with it.

I've put a demo of of the implementation as described in this comment at https://bfo.com/publisher/?saved/18c0fb7cb6c548f5#svg

**Comment by astearns** [comment]
Having shape-margin work exactly the same between floats and initial-letter seems like a good result to me.

**Comment by fantasai** [comment]
@astearns / @faceless2 Mind reviewing the edits to make sure I got them right? https://github.com/w3c/csswg-drafts/commit/6b1e2a2a23af140cff5edff2a8271700b8c8b9e6

**Comment by faceless2** [comment]
That change LGTM, thanks!

**Comment by astearns** [comment]
@fantasai I'd put a paragraph break before "In both cases." Making it part of the second paragraph of the definition obscures the fact that it applies to the first paragraph as well.

**Comment by fantasai** [comment]
@astearns OK, done! Thanks for the review~
